codegen/masm/src/codegen.rs

```diff
@@ -0,0 +1,273 @@
+                        let scalar = vec.get(*pos).ok_or(CodegenError::InvalidAccessType)?;
+                        self.ins.push(format!("push.{scalar}").into());
+                    }
+                    (ConstantValueExpr::Matrix(mat), AccessType::Matrix(x, y)) => {
+                        let scalar = mat
+                            .get(*x)
+                            .and_then(|v| v.get(*y))
+                            .ok_or(CodegenError::InvalidAccessType)?;
+                        self.ins.push(format!("push.{scalar}").into());
+                    }
+                    _ => return Err(CodegenError::InvalidAccessType),
+                },
+                Entry::Vacant(_) => return Err(CodegenError::UnknowConstant),
+            },
+            Value::InlineConstant(value) => {
+                self.ins.push(format!("push.{value}").into());
+            }
+            Value::TraceElement(access) => {
+                // eventually larger offsets will be supported
+                if access.row_offset() > 1 {
+                    return Err(CodegenError::InvalidRowOffset);
+                }
+
+                // should always be one
+                if access.size() != 1 {
+                    return Err(CodegenError::InvalidSize);
+                }
+
+                // encoding defined at https://github.com/0xPolygonMiden/miden-vm/issues/875
+                // TODO: handle ext2
+                let words: u64 = (access.col_idx() / 4)
+                    .try_into()
+                    .map_err(|_| CodegenError::InvalidColumnIndex)?;
+                let address = self.config.ood_frame_address + words;
+                self.ins.push("padw".into());
+                self.ins.push(format!("mem_loadw.{address}").into());
+
+                match (access.trace_segment(), access.col_idx() % 4) {
+                    (MAIN_TRACE, 0) => {
+                        self.ins.push("movdn.3".into());
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                    }
+                    (MAIN_TRACE, 1) => {
+                        self.ins.push("drop".into());
+                        self.ins.push("movdn.2".into());
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                    }
+                    (MAIN_TRACE, 2) => {
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                        self.ins.push("swap".into());
+                        self.ins.push("drop".into());
+                    }
+                    (MAIN_TRACE, 3) => {
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                        self.ins.push("drop".into());
+                    }
+                    _ => {
```

Comment:
question: is every trace besides `main` using `ext2`? (as in, should I use `(AUX_TRACE, ...)` or `(_, ...)` when implementing ext2 support?)

Reply:
All elements in the OOD frame are in the extension field (including the `main` segment values).

Reply:
Ah, I misunderstood that. Thanks for the clarification!

Reply:
code is updated to use quadratic extension fields